### PR TITLE
Rule 3.14.2 string template format

### LINKED ISSUE #301: Rule 3.14.2: String template format
### <a name="r3.8"></a>Rule 3.14.2: String template format
Redundant curly braces in string templates

In String templates there should not be redundant curly braces. In case of using a not complex statement (one argument)
 there should not be curly braces.

Bad example:
```kotlin
val someString = "${myArgument} ${myArgument.foo()}"
```

Valid example:
```kotlin
val someString = "$myArgument ${myArgument.foo()}"
```

Redundant string template
In case string template contains only one variable - there is no need to use string template. Use this variable directly.

Bad example:
```kotlin
val someString = "$myArgument"
```

Valid example:
```kotlin
val someString = myArgument
```

### PR DESCRIPTION
**Pull request by aktsay6**
### What's done:
 * Added rule logic
 * Added warn and fix test

## Actions checklist
* [X] Implemented Rule, added Warnings
* [X] Added tests on checks
* [X] Added tests on fixers
* [X] Updated diktat-analysis.yml
* [X] Updated available-rules.md

Closes #301

### CODE REVIEW

File: diktat-analysis.yml [hunk 1]
```diff
@@ -235,4 +235,10 @@
 - name: TYPE_ALIAS
   enabled: true
   configuration:
-    typeReferenceLength: '25' # max length of type reference
\ No newline at end of file
+    typeReferenceLength: '25' # max length of type reference
+- name: REDUNDANT_CURLY_BRACES
```
Comment (akuleshov7) [inline diktat-analysis.yml:239]: add description
change names: STRING_TEMPLATE_CURLY_BRACES
STRING_TEMPLATE_QUOTES

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringTemplateFormatRule.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                       autoCorrect: Boolean,
+                       emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit) {
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == LONG_STRING_TEMPLATE_ENTRY) {
+            handleLongStringTemplate(node)
+        }
+
+        if (node.elementType == SHORT_STRING_TEMPLATE_ENTRY) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringTemplateFormatRule.kt:36]: Please add comment about why these two `if`s cannot be combined in `if-else`

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringTemplateFormatRule.kt [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import org.cqfn.diktat.common.config.rules.RulesConfig
+import org.cqfn.diktat.ruleset.constants.Warnings
+import org.cqfn.diktat.ruleset.utils.findAllNodesWithSpecificType
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.CompositeElement
+import org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.LeafPsiElement
+
+/**
+ * In String templates there should not be redundant curly braces. In case of using a not complex statement (one argument)
+ * there should not be curly braces.
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringTemplateFormatRule.kt:20]: Please take a look at [this comment](https://github.com/cqfn/diKTat/issues/301#issuecomment-697280825). Please add it as fixme, I think we can't do anything with this now.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringTemplateFormatRule.kt [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun handleLongStringTemplate(node: ASTNode) {
+        // Checking if in long templates {a.foo()} there are function calls or class toString call
+        if (node.findAllNodesWithSpecificType(COLONCOLON).isEmpty()
+                && node.findAllNodesWithSpecificType(DOT_QUALIFIED_EXPRESSION).isEmpty()) {
+            Warnings.STRING_TEMPLATE_CURLY_BRACES.warnAndFix(configRules, emitWarn, isFixMode, node.text, node.startOffset, node) {
+                val identifierName = node.findChildByType(REFERENCE_EXPRESSION)!!.text
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringTemplateFormatRule.kt:44]: Will this branch trigger for strings like `"${1.0}"`? There can be expressions that don't have references.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringTemplateFormatRule.kt [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                shortTemplate.addChild(reference)
+                reference.addChild(LeafPsiElement(IDENTIFIER, identifierName))
+                node.treeParent.removeChild(node)
+            }
+        }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun handleShortStringTemplate(node: ASTNode) {
+        val identifierName = node.findChildByType(REFERENCE_EXPRESSION)!!.text
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringTemplateFormatRule.kt:59]: Here I can't think of an example where there is no reference, but to be safe please add handling of null.

**Review by petertrr — CHANGES_REQUESTED** [review]

**Review by petertrr — APPROVED** [review]
LGTM

**Pull request merged**